mava/wrappers/jumanji.py

```diff
@@ -265,3 +272,124 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
+        env.num_agents = env._num_vehicles
+        env.time_limit = env._num_customers + 1  # added for consistency
+        env.action_dim = env._num_customers + 1  # n_costumers + 1 starter node
+        super().__init__(env, False)
+        self.has_global_state = add_global_state
+        self._env = env
+
+    def modify_timestep(self, timestep: TimeStep, state: State) -> TimeStep[Observation]:
+        observation, global_observation = self._flatten_observation(timestep.observation)
+        obs_data = {
+            "agents_view": observation,
+            "action_mask": timestep.observation.action_mask,
+            "step_count": jnp.repeat(state.step_count, (self.num_agents)),
+        }
+        if self.has_global_state:
+            obs_data["global_state"] = global_observation
+            observation = ObservationGlobalState(**obs_data)
+        else:
+            observation = Observation(**obs_data)
+
+        reward = jnp.repeat(timestep.reward, (self.num_agents))
+        discount = jnp.repeat(timestep.discount, (self.num_agents))
+        return timestep.replace(observation=observation, reward=reward, discount=discount)
+
+    def _flatten_observation(
+        self, observation: MultiCvrpObservation
+    ) -> Tuple[chex.Array, Union[None, chex.Array]]:
+        """
+        Concatenates all observation fields into a single array.
+
+        Args:
+            observation (MultiCvrpObservation): The raw observation NamedTuple provided by jumanji.
+
+        Returns:
+            observations (chex.Array): Concatenated individual observations for each agent,
+                shaped (num_agents, vehicle_info + customer_info).
+            global_observation (Union[None, chex.Array]): Concatenated global observation
+                shaped (num_agents, global_info) if has_global_state = True, None otherwise.
+        """
+        global_observation = None
+        # N: number of nodes, same as _num_customers + 1
+        # V: number of vehicles, same as num_agents
+        # nodes are composed of (x, y, demands)
+        # Windows are composed of (start_time, end_time)
+        # Coeffs are composed of (early, late)
+        # Vehicles have ((x, y), local_time, capacity)
+
+        # Tuple[(N, 3), (N, 2), (N, 2)]
+        customers_info, _ = tree_util.tree_flatten(
+            (observation.nodes, observation.windows, observation.coeffs)
+        )
+        # Tuple[(V, 2), (V, 1), (V, 1)]
+        vehicles_info, _ = tree_util.tree_flatten(observation.vehicles)
+
+        # (N * 7, )
+        customers_info = jnp.column_stack(customers_info).ravel()
+        # (V, 4)
+        vehicles_info = jnp.column_stack(vehicles_info)
+
+        if self.has_global_state:
+            # (V * 4 + N * 7, )
+            global_observation = jnp.concatenate((vehicles_info.ravel(), customers_info))
+            # (V, N * 7 + V * 4)
+            global_observation = jnp.tile(global_observation, (self.num_agents, 1))
+
+        # (V, N * 7)
```

Comment:
Just a general note, it is quite hard to follow where the 4 and 7 in the shapes come from. Is it possible to please elaborate a bit more in the comments?